internal/controllers/gateway/gateway_controller.go

```diff
@@ -0,0 +1,594 @@
+	// mode.
+	//
+	// NOTE: it's currently required that this annotation be present on all Gateway
+	// resources: "pre-provisioned" mode is the only supported mode at this time.
+	PreProvisionedAnnotation = "konghq.com/gateway-pre-provisioned-mode"
+
+	// ProvisionedGatewaysUnsupported is an error used whenever a failure occurs
+	// due to a Gateway that is not properly configured for pre-provisioned mode.
+	ProvisionedGatewaysUnsupported = fmt.Errorf("invalid gateway spec: pre-provisioned gateways are not currently supported") //nolint:revive
+)
+
+// -----------------------------------------------------------------------------
+// Gateway Controller - GatewayReconciler
+// -----------------------------------------------------------------------------
+
+// GatewayReconciler reconciles a Gateway object
+type GatewayReconciler struct { //nolint:revive
+	client.Client
+
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	Proxy           proxy.Proxy
+	PublishService  string
+	WatchNamespaces []string
+
+	allowedRoutes     gatewayv1alpha2.AllowedRoutes
+	publishServiceRef types.NamespacedName
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// determine the AllowedRoutes for Gateways
+	group := gatewayv1alpha2.Group(gatewayv1alpha2.GroupName)
+	r.allowedRoutes.Kinds = []gatewayv1alpha2.RouteGroupKind{{
+		Group: &group,
+		Kind:  gatewayv1alpha2.Kind("HTTPRoute"), // currently ONLY HTTPRoute is supported
+	}}
+
+	// determine whether AllowedRoutes will come from ALL namespaces, or specific ones
+	namespaces := gatewayv1alpha2.RouteNamespaces{}
+	if len(r.WatchNamespaces) < 1 {
+		fromAll := gatewayv1alpha2.NamespacesFromAll
+		namespaces.From = &fromAll
+	} else {
+		fromSelector := gatewayv1alpha2.NamespacesFromSelector
+		namespaces.From = &fromSelector
+	}
+	r.allowedRoutes.Namespaces = &namespaces
+
+	// convert the service ref into a NamespacedName
+	publishServiceSplit := strings.Split(r.PublishService, "/")
+	if len(publishServiceSplit) != 2 {
+		return fmt.Errorf("--publish-service expected in format 'namespace/name' but got %s", r.PublishService)
+	}
+	r.publishServiceRef = types.NamespacedName{
+		Namespace: publishServiceSplit[0],
+		Name:      publishServiceSplit[1],
+	}
+
+	// generate the controller object and attach it to the manager and link the reconciler object
+	c, err := controller.New("gateway-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// watch Gateway objects, filtering out any Gateways which are not configured with
+	// a supported GatewayClass controller name.
+	if err := c.Watch(
+		&source.Kind{Type: &gatewayv1alpha2.Gateway{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.gatewayHasMatchingGatewayClass),
+	); err != nil {
+		return err
+	}
+
+	// watch for updates to gatewayclasses, if any gateway classes change, enqueue
```

Comment:
Do we eventually end up with a separate GatewayClass reconciler? I don't think we have anything else where a change in one resource necessitates an update in another, which seems necessary here--the others aren't like this, but because they don't have to be. Proxy Service and Status is a sort-of exception but it's handled elsewhere.

Reply:
> This is a significant departure from the model where one reconciler watches one kind

This is a common implementation strategy among our neighbor gateway implementations to ensure that updates on Gateways occur in response to GatewayClass changes.

> Is this intentional?

As per the comment explaining why this is being done here, yes this is intentional.

> Do we eventually end up with a separate GatewayClass reconciler?

In time we most likely end up with a separate GatewayClass reconciler as our implementation grows, but at least for the purposes of this iteration it's not yet needed to start integrating webhook validation and `HTTPRoute`.

Reply:
We discussed this further in our zoom today, let me know if there's any more follow up needed here.